File: examples/python/buffers.py
#!/usr/bin/env python3

#####################################################################
# This script presents different buffers and formats.
# OpenCV is used here to display images, install it or remove any
# references to cv2
# Configuration is loaded from "../../scenarios/basic.cfg" file.
# <episodes> number of episodes are played.
# Random combination of buttons is chosen for every action.
#
# To see the scenario description go to "../../scenarios/README.md"
#####################################################################

from __future__ import print_function

from random import choice
import vizdoom as vzd
from argparse import ArgumentParser

DEFAULT_CONFIG = "../../scenarios/deadly_corridor.cfg"

import cv2

if __name__ == "__main__":

    parser = ArgumentParser("ViZDoom example showing different buffers (screen, depth, labels).")
    parser.add_argument(dest="config",
                        default=DEFAULT_CONFIG,
                        nargs="?",
                        help="Path to the configuration file of the scenario."
                             " Please see "
                             "../../scenarios/*cfg for more scenarios.")

    args = parser.parse_args()

    game = vzd.DoomGame()

    # Use other config file if you wish.
    game.load_config(args.config)

    # game.set_console_enabled(True)
    # game.set_window_visible(False)

    # Just umcomment desired format for screen buffer (and map buffer).
    # The last uncommented will be applied.
    # Formats with C (CRCGCB and CBCGCR) were omitted cause they are not cv2 friendly.
    # Default format is ScreenFormat.CRCGCB.

    # OpenCV uses a BGR colorspace by default.
    game.set_screen_format(vzd.ScreenFormat.BGR24)

    # game.set_screen_format(ScreenFormat.RGB24)
    # game.set_screen_format(ScreenFormat.RGBA32)
    # game.set_screen_format(ScreenFormat.ARGB32)
    # game.set_screen_format(ScreenFormat.BGRA32)
    # game.set_screen_format(ScreenFormat.ABGR32)
    # game.set_screen_format(ScreenFormat.GRAY8)

    # Raw Doom buffer with palette's values. This one makes no sense in particular
    # game.set_screen_format(ScreenFormat.DOOM_256_COLORS)

    # Sets resolution for all buffers.
    game.set_screen_resolution(vzd.ScreenResolution.RES_640X480)

    # Enables depth buffer.
    game.set_depth_buffer_enabled(True)

    # Enables labeling of in game objects labeling.
    game.set_labels_buffer_enabled(True)

    # Enables buffer with top down map of he current episode/level .
    game.set_automap_buffer_enabled(True)
    game.set_automap_mode(vzd.AutomapMode.OBJECTS)
    game.set_automap_rotate(False)
    game.set_automap_render_textures(False)

    game.set_render_hud(True)
    game.set_render_minimal_hud(False)

    game.set_mode(vzd.Mode.SPECTATOR)
    game.init()

    actions = [[True, False, False], [False, True, False], [False, False, True]]

    episodes = 10
    sleep_time = 0.028

    for i in range(episodes):
        print("Episode #" + str(i + 1))

        # Not needed for the first episode but the loop is nicer.
        game.new_episode()
        while not game.is_episode_finished():
            # Gets the state and possibly do something with it
            state = game.get_state()

            # Display all the buffers here!

            # Just screen buffer, given in selected format. This buffer is always available.
            screen = state.screen_buffer
            cv2.imshow('ViZDoom Screen Buffer', screen)

            # Depth buffer, always in 8-bit gray channel format.
            # This is most fun. It looks best if you inverse colors.
            depth = state.depth_buffer
            if depth is not None:
                cv2.imshow('ViZDoom Depth Buffer', depth)

            # Labels buffer, always in 8-bit gray channel format.
            # Shows only visible game objects (enemies, pickups, exploding barrels etc.), each with unique label.
            # Labels data are available in state.labels, also see labels.py example.
            labels = state.labels_buffer
            if labels is not None:
                cv2.imshow('ViZDoom Labels Buffer', labels)

            # Map buffer, in the same format as screen buffer.
            # Shows top down map of the current episode/level.
            automap = state.automap_buffer
            if automap is not None:
                cv2.imshow('ViZDoom Map Buffer', automap)

            cv2.waitKey(int(sleep_time * 1000))

            game.make_action(choice(actions))

            print("State #" + str(state.number))
            print("=====================")

        print("Episode finished!")
        print("************************")

    cv2.destroyAllWindows()

File: examples/lua/buffers.lua
#!/usr/bin/env th

----------------------------------------------------------------------
-- This script presents SPECTATOR mode. In SPECTATOR mode you play and
-- your agent can learn from it.
-- Configuration is loaded from "../../scenarios/<SCENARIO_NAME>.cfg" file.
--
-- To see the scenario description go to "../../scenarios/README.md"
----------------------------------------------------------------------

require "vizdoom"
require "torch"
require "image"
require "sys"

game = vizdoom.DoomGame()

-- Choose scenario config file you wish to watch.
-- Don't load two configs cause the second will overrite the first one.
-- Multiple config files are ok but combining these ones doesn't make much sense.

--game:loadConfig("../../scenarios/basic.cfg")
game:loadConfig("../../scenarios/deadly_corridor.cfg")
--game:loadConfig("../../scenarios/deathmatch.cfg")
--game:loadConfig("../../scenarios/defend_the_center.cfg")
--game:loadConfig("../../scenarios/defend_the_line.cfg")
--game:loadConfig("../../scenarios/health_gathering.cfg")
--game:loadConfig("../../scenarios/my_way_home.cfg")
--game:loadConfig("../../scenarios/predict_position.cfg")
--game:loadConfig("../../scenarios/take_cover.cfg")

-- Enables freelook in engine
game:addGameArgs("+freelook 1")

game:setScreenResolution(vizdoom.ScreenResolution.RES_640X480)
game:setWindowVisible(true)

-- Enables labeling of the in game objects.
game:setLabelsBufferEnabled(true)

-- Enables depth buffer.
game:setDepthBufferEnabled(true)

-- Enables labeling of in game objects labeling.
game:setLabelsBufferEnabled(true)

-- Enables buffer with top down map of he current episode/level .
game:setAutomapBufferEnabled(true)
game:setAutomapMode(vizdoom.AutomapMode.OBJECTS)
game:setAutomapRotate(false)
game:setAutomapRenderTextures(false)

game:setRenderHud(true)
game:setRenderMinimalHud(false)

game:init()

local actions = {
    [1] = torch.IntTensor({1,0,0}),
    [2] = torch.IntTensor({0,1,0}),
    [3] = torch.IntTensor({0,0,1})
}

local episodes = 10
local sleepTime = 0.028

for i = 1, episodes do

    print("Episode #" .. i)

    game:newEpisode()

    while not game:isEpisodeFinished() do

        -- Gets the state.
        local state = game:getState()

        local screen = state.screenBuffer
        screenWin = image.display{image=screen, legend="ViZDoom Screen Buffer", offscreen=false, win=screenWin}

        -- Depth buffer, always in 8-bit gray channel format.
        -- This is most fun. It looks best if you inverse colors.
        local depth = state.depthBuffer
        if depth then
            depthWin = image.display{image=depth, legend="ViZDoom Depth Buffer", offscreen=false, win=depthWin}
        end

        -- Labels buffer, always in 8-bit gray channel format.
        -- Shows only visible game objects (enemies, pickups, exploding barrels etc.), each with unique label.
        -- Labels data are available in state.labels, also see labels.py example.
        local labels = state.labelsBuffer
        if labels then
            labelsWin = image.display{image=labels, legend="ViZDoom Labels Buffer", offscreen=false, win=labelsWin}
        end

        -- Map buffer, in the same format as screen buffer.
        -- Shows top down map of the current episode/level.
        local automap = state.automapBuffer
        if automap then
            automapWin = image.display{image=automap, legend="ViZDoom Automap Buffer", offscreen=false, win=automapWin}
        end

        local action = actions[torch.random(#actions)]
        local reward = game:makeAction(action)

        print("State #" .. state.number)
        print("=====================")

        if sleepTime > 0 then
            sys.sleep(sleepTime)
        end

    end

    print("Episode finished.")
    print("total reward: " .. game:getTotalReward())
    print("************************")

end

game:close()
